File: src/temporal_xray/temporal_client.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from typing import Any

from temporalio.client import Client, TLSConfig


@dataclass
class TemporalConfig:
    address: str = ""
    namespace: str = "default"
    tls_cert_path: str | None = None
    tls_key_path: str | None = None
    api_key: str | None = None


_connection_client: Client | None = None
_client_cache: dict[str, Client] = {}
_active_config: TemporalConfig | None = None


def _get_config() -> TemporalConfig:
    if _active_config:
        return _active_config
    return TemporalConfig(
        address=os.environ.get("TEMPORAL_ADDRESS", "localhost:7233"),
        namespace=os.environ.get("TEMPORAL_NAMESPACE", "default"),
        tls_cert_path=os.environ.get("TEMPORAL_TLS_CERT_PATH") or None,
        tls_key_path=os.environ.get("TEMPORAL_TLS_KEY_PATH") or None,
        api_key=os.environ.get("TEMPORAL_API_KEY") or None,
    )


def _build_tls_config(config: TemporalConfig) -> TLSConfig | None:
    if not config.tls_cert_path and not config.tls_key_path and not config.api_key:
        return None

    if config.api_key:
        # API key auth (Temporal Cloud) — TLS required but certs handled by cloud
        return TLSConfig()

    if config.tls_cert_path and config.tls_key_path:
        with open(config.tls_cert_path, "rb") as f:
            client_cert = f.read()
        with open(config.tls_key_path, "rb") as f:
            client_key = f.read()
        return TLSConfig(client_cert=client_cert, client_private_key=client_key)

    return None


async def _ensure_connection(config: TemporalConfig | None = None) -> Client:
    global _connection_client

    if _connection_client is not None:
        return _connection_client

    cfg = config or _get_config()
    tls = _build_tls_config(cfg)

    try:
        _connection_client = await Client.connect(
            cfg.address,
            namespace=cfg.namespace,
            tls=tls or False,
            api_key=cfg.api_key,
        )
        return _connection_client
    except Exception as e:
        raise RuntimeError(
            f"Cannot connect to Temporal server at {cfg.address}. "
            f"Verify TEMPORAL_ADDRESS is correct and the server is running. Details: {e}"
        ) from e


async def get_temporal_client(namespace: str | None = None) -> Client:
    """Get a Temporal client for the given namespace.

    If no namespace is provided, uses the configured default.
    Clients are cached per namespace.
    """
    config = _get_config()
    effective_namespace = namespace or config.namespace

    cached = _client_cache.get(effective_namespace)
    if cached is not None:
        return cached

    # For API key auth with a non-default namespace, create a separate client
    if config.api_key and effective_namespace != config.namespace and _connection_client is not None:
        tls = _build_tls_config(config)
        client = await Client.connect(
            config.address,
            namespace=effective_namespace,
            tls=tls or False,
            api_key=config.api_key,
        )
        _client_cache[effective_namespace] = client
        return client

    client = await _ensure_connection()

    # If requesting a different namespace, create a new client for it
    if effective_namespace != config.namespace:
        tls = _build_tls_config(config)
        client = await Client.connect(
            config.address,
            namespace=effective_namespace,
            tls=tls or False,
            api_key=config.api_key,
        )

    _client_cache[effective_namespace] = client
    return client


async def connect_to_server(
    config: TemporalConfig,
) -> dict[str, Any]:
    """Connect to a different Temporal server."""
    global _active_config
    await close_connection()
    _active_config = config
    await get_temporal_client()
    return get_connection_status()


def get_connection_status() -> dict[str, Any]:
    """Get the current connection status."""
    config = _get_config()
    if config.api_key:
        auth_type = "api_key"
    elif config.tls_cert_path:
        auth_type = "mtls"
    else:
        auth_type = "none"

    return {
        "address": config.address,
        "namespace": config.namespace,
        "authType": auth_type,
        "connected": _connection_client is not None,
    }


async def close_connection() -> None:
    """Close the connection and clear caches."""
    global _connection_client
    _client_cache.clear()
    if _connection_client is not None:
        await _connection_client.service_client.disconnect()
        _connection_client = None


# --- Payload Decoding ---


DecodedPayload = (
    dict[str, Any] | list[Any] | str | int | float | bool | None
)


def decode_payload(
    payload: Any | None,
    truncate_at: int | None = None,
) -> DecodedPayload:
    """Decode a Temporal payload into a usable value."""
    if payload is None or not hasattr(payload, "data") or payload.data is None:
        return None

    encoding = ""
    if hasattr(payload, "metadata") and payload.metadata:
        enc_bytes = payload.metadata.get("encoding")
        if enc_bytes:
            encoding = enc_bytes.decode("utf-8") if isinstance(enc_bytes, (bytes, bytearray)) else str(enc_bytes)

    # Encrypted payloads
    if encoding in ("binary/encrypted", "encoding/encrypted"):
        return {
            "_type": "encrypted",
            "note": "Payloads are encrypted. Configure a codec server endpoint for decryption.",
        }

    # Protobuf payloads
    if encoding in ("binary/protobuf",) or encoding.startswith("json/protobuf"):
        message_type = "unknown"
        if hasattr(payload, "metadata") and payload.metadata:
            mt_bytes = payload.metadata.get("messageType")
            if mt_bytes:
                message_type = mt_bytes.decode("utf-8") if isinstance(mt_bytes, (bytes, bytearray)) else str(mt_bytes)
        return {
            "_type": "protobuf",
            "messageType": message_type,
            "note": "Install the project's protobuf definitions for full deserialization",
        }

    # JSON payloads (default)
    try:
        data = payload.data
        if isinstance(data, (bytes, bytearray)):
            raw = data.decode("utf-8")
        else:
            raw = str(data)

        if truncate_at and len(data) > truncate_at:
            return {
                "_truncated": True,
                "preview": raw[:500],
                "fullSizeBytes": len(data),
            }

        return json.loads(raw)
    except (json.JSONDecodeError, UnicodeDecodeError):
        if isinstance(data, (bytes, bytearray)):
            return data.decode("utf-8", errors="replace")
        return str(data)


def decode_payloads(
    payloads: Any | None,
    truncate_at: int | None = None,
) -> list[Any]:
    """Decode a list of Temporal payloads."""
    if not payloads:
        return []
    return [decode_payload(p, truncate_at) for p in payloads]


def decode_single_payload(
    payloads: Any | None,
    truncate_at: int | None = None,
) -> Any:
    """Unwrap a single-element payload array to a single value."""
    decoded = decode_payloads(payloads, truncate_at)
    if len(decoded) == 1:
        return decoded[0]
    if len(decoded) == 0:
        return None
    return decoded

File: src/temporal-client.ts
import { Connection, Client } from "@temporalio/client";
import type { DecodedPayload } from "./types.js";

const textDecoder = new TextDecoder();

let connectionInstance: Connection | null = null;
const clientCache = new Map<string, Client>();

export interface TemporalConfig {
  address: string;
  namespace: string;
  tlsCertPath?: string;
  tlsKeyPath?: string;
  apiKey?: string;
}

let activeConfig: TemporalConfig | null = null;

function getConfig(): TemporalConfig {
  if (activeConfig) return activeConfig;
  return {
    address: process.env.TEMPORAL_ADDRESS || "localhost:7233",
    namespace: process.env.TEMPORAL_NAMESPACE || "default",
    tlsCertPath: process.env.TEMPORAL_TLS_CERT_PATH || undefined,
    tlsKeyPath: process.env.TEMPORAL_TLS_KEY_PATH || undefined,
    apiKey: process.env.TEMPORAL_API_KEY || undefined,
  };
}

async function buildTlsConfig(
  config: TemporalConfig
): Promise<import("@temporalio/client").TLSConfig | undefined> {
  if (!config.tlsCertPath && !config.tlsKeyPath && !config.apiKey) {
    return undefined;
  }

  if (config.apiKey) {
    // API key auth (Temporal Cloud) — TLS is required but certs are handled by the cloud
    return {};
  }

  if (config.tlsCertPath && config.tlsKeyPath) {
    const fs = await import("fs");
    return {
      clientCertPair: {
        crt: fs.readFileSync(config.tlsCertPath),
        key: fs.readFileSync(config.tlsKeyPath),
      },
    };
  }

  return undefined;
}

async function ensureConnection(): Promise<Connection> {
  if (connectionInstance) return connectionInstance;

  const config = getConfig();
  const tls = await buildTlsConfig(config);

  const connectionOptions: import("@temporalio/client").ConnectionOptions = {
    address: config.address,
    tls,
  };

  // Add API key metadata if configured
  if (config.apiKey) {
    connectionOptions.metadata = {
      "temporal-namespace": config.namespace,
    };
    connectionOptions.apiKey = config.apiKey;
  }

  try {
    connectionInstance = await Connection.connect(connectionOptions);
    return connectionInstance;
  } catch (error) {
    const message =
      error instanceof Error ? error.message : String(error);
    throw new Error(
      `Cannot connect to Temporal server at ${config.address}. Verify TEMPORAL_ADDRESS is correct and the server is running. Details: ${message}`
    );
  }
}

/**
 * Get a Temporal client for the given namespace.
 * If no namespace is provided, uses the configured default.
 * Clients are cached per namespace and share the same connection.
 */
export async function getTemporalClient(
  namespace?: string
): Promise<Client> {
  const config = getConfig();
  const effectiveNamespace = namespace || config.namespace;

  const cached = clientCache.get(effectiveNamespace);
  if (cached) return cached;

  // For API key auth with a non-default namespace, we need a new connection
  // because the temporal-namespace header is set at the connection level
  if (
    config.apiKey &&
    effectiveNamespace !== config.namespace &&
    connectionInstance
  ) {
    const tls = await buildTlsConfig(config);
    const conn = await Connection.connect({
      address: config.address,
      tls,
      metadata: { "temporal-namespace": effectiveNamespace },
      apiKey: config.apiKey,
    });
    const client = new Client({ connection: conn, namespace: effectiveNamespace });
    clientCache.set(effectiveNamespace, client);
    return client;
  }

  const connection = await ensureConnection();
  const client = new Client({
    connection,
    namespace: effectiveNamespace,
  });
  clientCache.set(effectiveNamespace, client);
  return client;
}

/**
 * Connect to a different Temporal server. Closes the existing connection
 * and clears the client cache.
 */
export async function connectToServer(
  config: TemporalConfig
): Promise<{ address: string; namespace: string; authType: string }> {
  await closeConnection();
  activeConfig = config;

  // Verify the connection works
  await getTemporalClient();

  return getConnectionStatus();
}

/**
 * Get the current connection status.
 */
export function getConnectionStatus(): {
  address: string;
  namespace: string;
  authType: string;
  connected: boolean;
} {
  const config = getConfig();
  const authType = config.apiKey
    ? "api_key"
    : config.tlsCertPath
      ? "mtls"
      : "none";

  return {
    address: config.address,
    namespace: config.namespace,
    authType,
    connected: connectionInstance !== null,
  };
}

export async function closeConnection(): Promise<void> {
  clientCache.clear();
  if (connectionInstance) {
    await connectionInstance.close();
    connectionInstance = null;
  }
}

/**
 * Decode a Temporal payload into a usable value.
 * Handles JSON, protobuf, encrypted, and large payloads.
 */
export function decodePayload(
  payload: { data?: Uint8Array; metadata?: Record<string, Uint8Array> } | null | undefined,
  truncateAt?: number
): DecodedPayload {
  if (!payload || !payload.data) return null;

  const encoding = payload.metadata?.["encoding"]
    ? textDecoder.decode(payload.metadata["encoding"])
    : "";

  // Encrypted payloads
  if (encoding === "binary/encrypted" || encoding === "encoding/encrypted") {
    return {
      _type: "encrypted",
      note: "Payloads are encrypted. Configure a codec server endpoint for decryption.",
    };
  }

  // Protobuf payloads
  if (encoding === "binary/protobuf" || encoding.startsWith("json/protobuf")) {
    const messageType = payload.metadata?.["messageType"]
      ? textDecoder.decode(payload.metadata["messageType"])
      : "unknown";
    return {
      _type: "protobuf",
      messageType,
      note: "Install the project's protobuf definitions for full deserialization",
    };
  }

  // JSON payloads (default)
  try {
    const raw = textDecoder.decode(payload.data);

    // Large payload truncation
    if (truncateAt && payload.data.byteLength > truncateAt) {
      return {
        _truncated: true,
        preview: raw.slice(0, 500),
        fullSizeBytes: payload.data.byteLength,
      };
    }

    return JSON.parse(raw);
  } catch {
    // If not valid JSON, return as string
    return textDecoder.decode(payload.data);
  }
}

/**
 * Safely decode payloads from a Temporal Payload array.
 */
export function decodePayloads(
  payloads: Array<{ data?: Uint8Array; metadata?: Record<string, Uint8Array> }> | null | undefined,
  truncateAt?: number
): unknown[] {
  if (!payloads || payloads.length === 0) return [];
  return payloads.map((p) => decodePayload(p, truncateAt));
}

/**
 * Unwrap a single-element payload array to a single value.
 */
export function decodeSinglePayload(
  payloads: Array<{ data?: Uint8Array; metadata?: Record<string, Uint8Array> }> | null | undefined,
  truncateAt?: number
): unknown {
  const decoded = decodePayloads(payloads, truncateAt);
  return decoded.length === 1 ? decoded[0] : decoded.length === 0 ? null : decoded;
}
